### PR TITLE
Let vm_util be C-only, split C++ part into vm_util_cc

### PR DESCRIPTION
**Pull request by bjacob**
This seems to be a good chunk of the route to entirely C++-free binaries such as `iree-run-module,` which I believe will be a substantial binary size shrink, and will enable securing that with binary checks such as "`nm iree-run-module` says there is no `operator new`"

### CODE REVIEW

File: runtime/src/iree/tooling/BUILD [hunk 1]
```diff
@@ -136,28 +136,41 @@ iree_runtime_cc_test(
...
     name = "vm_util",
-    srcs = ["vm_util.cc"],
+    srcs = ["vm_util.c"],
     hdrs = ["vm_util.h"],
     deps = [
         ":numpy_io",
         "//runtime/src/iree/base",
-        "//runtime/src/iree/base:cc",
         "//runtime/src/iree/base:tracing",
         "//runtime/src/iree/base/internal:span",
```
Comment (benvanik) [inline runtime/src/iree/tooling/BUILD:145]: can drop span from here

File: runtime/src/iree/tooling/vm_util.h [hunk 2]
```diff
@@ -7,17 +7,16 @@
...
 #define IREE_TOOLING_VM_UTIL_H_
 
-#include <string>
-#include <vector>
-
 #include "iree/base/internal/span.h"
-#include "iree/base/status_cc.h"
+#include "iree/base/status.h"
 #include "iree/base/string_builder.h"
+#include "iree/base/string_view.h"
```
Comment (benvanik) [inline runtime/src/iree/tooling/vm_util.h:13]:
```suggestion
#include "iree/base/api.h"
```

File: runtime/src/iree/tooling/vm_util.h [hunk 3]
```diff
@@ -7,17 +7,16 @@
 #ifndef IREE_TOOLING_VM_UTIL_H_
 #define IREE_TOOLING_VM_UTIL_H_
 
-#include <string>
-#include <vector>
-
 #include "iree/base/internal/span.h"
```
Comment (benvanik) [inline runtime/src/iree/tooling/vm_util.h:10]: can drop span from here

**Review by benvanik — CHANGES_REQUESTED** [review]

**Comment by bjacob** [comment]
@benvanik the core changes are in the renamed `vm_util.cc -> vm_util.c` file ; due to the renaming (???) it is folded by default in the github view, so make sure you don't miss it.  There's nontrivial stuff there: the `do {...} while` loop growing the string buffer to write a hal buffer to; the `vm::ref` -> `iree_vm_ref` changes; the simply dropped `vm::ref` around the `iree_vm_list_t` (i wasn't sure about that one).  I've been assuming good test coverage here, otherwise my code might be buggy.

**Comment by benvanik** [comment]
oh, I didn't catch that - will look

File: runtime/src/iree/tooling/vm_util.c [hunk 4]
```diff
@@ -6,19 +6,16 @@
...
-#include <cstdio>
-#include <type_traits>
-#include <vector>
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
 
 #include "iree/base/api.h"
-#include "iree/base/status_cc.h"
+#include "iree/base/target_platform.h"
```
Comment (benvanik) [inline runtime/src/iree/tooling/vm_util.c:14]: this should be included by base/api.h

File: runtime/src/iree/tooling/vm_util.c [hunk 5]
```diff
@@ -43,16 +40,23 @@ static bool iree_file_is_eof(FILE* file, uint64_t file_length) {
 }
 #endif  // IREE_PLATFORM_*
 
-using namespace iree;
-
-namespace iree {
+iree_status_t iree_allocate_and_copy_cstring_from_view(
```
Comment (benvanik) [inline runtime/src/iree/tooling/vm_util.c:43]:
```suggestion
static iree_status_t iree_allocate_and_copy_cstring_from_view(
```

File: runtime/src/iree/tooling/vm_util.c [hunk 6]
```diff
@@ -242,29 +260,35 @@ Status ParseToVariantList(iree_hal_allocator_t* device_allocator,
...
+  *out_list = variant_list;
+  return iree_ok_status();
 }
 
 // Prints a buffer view with contents without a trailing newline.
-static iree_status_t PrintBufferView(iree_hal_buffer_view_t* buffer_view,
-                                     iree_host_size_t max_element_count,
-                                     iree_string_builder_t* builder) {
-  std::string result_str(4096, '\0');
+static iree_status_t iree_print_buffer_view(iree_hal_buffer_view_t* buffer_view,
```
Comment (benvanik) [inline runtime/src/iree/tooling/vm_util.c:272]: iree_tooling_append_buffer_view_string shows how to do this without looping - we should move that into iree/hal/buffer_view_util (there's already the format and fprint in there, so having the string builder variant would be good) - could be `iree_hal_buffer_view_append` or something

File: runtime/src/iree/tooling/vm_util.c [hunk 7]
```diff
@@ -310,44 +336,38 @@ static Status PrintVariant(iree_vm_variant_t variant, size_t max_element_count,
...
-Status PrintVariantList(iree_vm_list_t* variant_list, size_t max_element_count,
-                        std::string* out_string) {
-  iree_string_builder_t builder;
-  iree_string_builder_initialize(iree_allocator_system(), &builder);
-  IREE_RETURN_IF_ERROR(
-      PrintVariantList(variant_list, max_element_count, &builder));
-  out_string->assign(iree_string_builder_buffer(&builder),
-                     iree_string_builder_size(&builder));
-  iree_string_builder_deinitialize(&builder);
   return iree_ok_status();
```
Comment (benvanik) [inline runtime/src/iree/tooling/vm_util.c:355]: missing end zone here

File: runtime/src/iree/tooling/vm_util.c [hunk 8]
```diff
@@ -242,29 +260,35 @@ Status ParseToVariantList(iree_hal_allocator_t* device_allocator,
...
       }
-      IREE_RETURN_IF_ERROR(iree_vm_list_push_value(variant_list.get(), &val));
+      IREE_RETURN_AND_END_ZONE_IF_ERROR(
+          z0, iree_vm_list_push_value(variant_list, &val));
     }
   }
-  *out_list = variant_list.release();
-  return OkStatus();
+  *out_list = variant_list;
+  return iree_ok_status();
```
Comment (benvanik) [inline runtime/src/iree/tooling/vm_util.c:268]: missing end zone here (maybe other places too)

**Review by benvanik — CHANGES_REQUESTED** [review]
the code in vm_util.c is now leaky - malloc + RETURN_IF_ERROR will leak, anything that was using vm::ref and switched off it will also leak on return, etc

it may be worth doing this in a few stages - first do the basic C++ -> C things (remove use of auto, etc), then start switching the printing/parsing to use only string builders, and then deal with handling cleanup/refactoring functions in order to make cleanup easier (alloc + status = call that does stuff + unconditional free + return status is easy to read and usually ends up with a useful in-place inner function that can be reused)

File: runtime/src/iree/tooling/vm_util.c (hunk 6 above)
Comment (bjacob) [reply runtime/src/iree/tooling/vm_util.c]: done

**Comment by bjacob** [comment]
What with callers of the form (iree-run-module-main.cc):

```c++
  vm::ref<iree_vm_list_t> inputs;
  IREE_RETURN_IF_ERROR(ParseToVariantList(
      device_allocator,
      iree::span<const std::string>{FLAG_function_inputs.data(),
                                    FLAG_function_inputs.size()},
      host_allocator, &inputs));
```

are they now leaking the list or is the `vm::ref` destroying it on destruction?

**Comment by benvanik** [comment]
vm::ref will destroy it and it should be fine - the leaks mostly come from the C side where there's an `iree_vm_ref_t` that is not released after a failure (IREE_RETURN_IF_ERROR returns from the function, etc)

**Comment by bjacob** [comment]
Got it. How about now?

**Comment by benvanik** [comment]
Looking good! RE ftello -> ftell, I don't recall but on all 64-bit platforms is ftell guaranteed to return 64-bit types? (want to make sure we aren't breaking >2GB models on ARM/x86)

**Comment by bjacob** [comment]
`ftell` returns `long`.  Is that 32bit on any 64bit platform ?

**Comment by benvanik** [comment]
I think windows/mingw is the only place where it isn't that we care about, and that has the special path in the code for that
https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models
anyone using ILP32 only has 32-bit pointers and should be fine with it
so lgtm!

**Comment by bjacob** [comment]
Glad it's a verbal LGTM - Github wants a formal 'approval' though! ;-)

**Comment by benvanik** [comment]
was waiting for you to request review :P will look now

**Review by benvanik — APPROVED** [review]

**Pull request merged**